feat: add support for did:ethr signed/meta transactions

lleifermann opened the pull request:
This PR adds support for sending transactions on behalf of another did:ethr.  

This is archived via updates to the underlying dependencies https://github.com/uport-project/ethr-did & https://github.com/decentralized-identity/ethr-did-resolver. 

To keep abstraction intact, this behavior may be triggered by passing a `metaIdentifier` to a didManager action which is picked up by the `did-provider-ethr`. If a metaIdentifier is present the underlying controller is passed this identifier as a signer which then signs and pays for the transactions. 

The methods `addKey` & `addService` now detect this presence as well and prepare the transaction differently then. 
Overall we tried to keep changes to other plugins as minimal as possible. 

Some todo'ish items: 

- Right now only addKey and addService have meta capabilities. `changeOwner` and `addDelegate` are NOT implemented yet. @mirceanis - Not sure how to tackle this with the current AbstractIdentifierProvider. We would have to put these actions into the `updateIdentifier` method. But with the current abstraction the did-provider-ethr would receive a didDocument there. Not sure how to proceed here with a good approach. 
We could change the abstract signature of `updateIdentifier` to something more easily hand-able (like actions being passed down or something). 

- We had to introduce a new 'algorithm' to the kms-local provider. All other existing options for creating eth signatures either modify the input data, or hash it again. We just need clean digest with the `r, s and v` params easily accessible afterwards. 

- The tests for the did-provider-ethr changes will fail as i am awaiting an update to ethr-did which fixes an encoding bug we found.

- After some testing and fiddling around we also noticed that if this package does not pass any options down to did-ethr some defaults for transactions will be chosen, which could result in the transaction failing. Gaslimit and Gasprice are always overridden as defaults. AFAIK if etherjs is not passed anything it will make a best guess for determining these parameters.

mirceanis commented on packages/kms-local/src/key-management-system.ts:
```diff
@@ -130,6 +130,8 @@ export class KeyManagementSystem extends AbstractKeyManagementSystem {
         return await this.eth_signMessage(managedKey.privateKeyHex, data)
       } else if (['eth_signTypedData', 'EthereumEip712Signature2021'].includes(algorithm)) {
         return await this.eth_signTypedData(managedKey.privateKeyHex, data)
+      } else if (['eth_rawSign'].includes(algorithm)) {
```
This KMS should also advertise this algorithm on the `meta.algorithms` property of the keys it creates.
Please see the `asManagedKeyInfo` method below.

At some point we will want to have finer control over the KMS chosen for particular operations based on the key capabilities (`meta.algorithms`), so I think it's worth keeping these 2 in sync.

mirceanis reviewed (changes requested):
Thanks for posting this!
Some tiny changes required regarding `eth_rawSign` and the unused `ethers` dependency.

---

On a broader note, from what I can see, the expectation here is that the Veramo instance is managing both the DID controller key and the transaction signing key. This does reduce some friction with regards to funding every controller key, but the real power of meta-transactions and of our modular key-management/did-management architecture would come from splitting these 2 actions (of signing the change and of sending the transaction) completely.

This would mean that `didManagerAddKey()` with a "meta-transaction" flag, would not actually send a transaction, but only perform the DID-controller signing. The result of this operation would then be sent to a transaction-signing service, which would be a separate thing.

This service could, of course, be implemented by the same Veramo instance, but it could also be a third party service.

I'm not requesting this change be added here, only flagging this as a future improvement.
Let me know what you think.

lleifermann commented:
@mirceanis i removed the ethers dependency and moved the ethr_rawSign into it's own method and is also now advertised in the meta.algorithm array for Secp256k1 keys.

Regarding the signature flow we wanted to keep it as straight forward as possible. But i see your point in separating these things. 

Whats your take on the missing `changeOwner` and `addDelegate` functions? Should we place those in the `updateIdentifier` function? AFAIK this would require a small change in the Abstract IdentifierProvider as it right now passes passes a DidDocument down. We figured a key/value map of possible `UpdateActions` would fit better as the concrete implementation of the provider can then expose them properly. (did-provider-ethr providing changeOwnerAction and addDelegateAction for example). Other providers then may provide other actions which fit in the `updateIdentifier` action.

With the current implementation we would need to do some delta comparison i believe (current did doc vs new did doc = wanted change)

mirceanis commented:
> @mirceanis i removed the ethers dependency and moved the ethr_rawSign into it's own method and is also now advertised in the meta.algorithm array for Secp256k1 keys.

That's great, thanks!

> Regarding the signature flow we wanted to keep it as straight forward as possible. But i see your point in separating these things.

It makes sense. What you did here is very valuable too, even if the full power of meta-tx is not used yet.

> Whats your take on the missing `changeOwner` and `addDelegate` functions? Should we place those in the `updateIdentifier` function? AFAIK this would require a small change in the Abstract IdentifierProvider as it right now passes passes a DidDocument down. We figured a key/value map of possible `UpdateActions` would fit better as the concrete implementation of the provider can then expose them properly. (did-provider-ethr providing changeOwnerAction and addDelegateAction for example). Other providers then may provide other actions which fit in the `updateIdentifier` action.
> 
> With the current implementation we would need to do some delta comparison i believe (current did doc vs new did doc = wanted change)

This is a tricky problem, because a change in `AbstractIdentifierProvider` would have to make sense for most DID methods (or at least more than the one we're dealing with now). My gut reaction is that custom control logic, that only makes sense to a certain DID method should be built as its own plugin and not try to morph the `DIDManager` API, but DID controller updates are probably common?
We started discussing this issue in #583 (wow, long time ago!), so we can continue the chat about this there.

mirceanis commented on __tests__/shared/ethrDidFlowSigned.ts:
```diff
@@ -82,13 +82,19 @@ export default (testContext: {
       const resolution = await agent.resolveDid({ didUrl: alice.did })
 
-      expect(resolution?.didDocument?.service?.[0].serviceEndpoint).toEqual(`http://localhost:${listeningPort}/foobar`)
+      expect(resolution.didDocument).toEqual(expect.objectContaining({
```
TIL about arrayContaining/objectContaining :)
This looks very clean

lleifermann commented:
Only thing missing here is now the failing import of the `ethr-did-controller`. Once the downstream bundling of `ethr-did` and `ethr-did-controller` is fixed i can include the version here. Then everything should be good to go as far as i can tell. 

For the addDelegate and changeOwner stuff i would propose a solution in #583

mirceanis commented on __tests__/shared/keyManager.ts:
```diff
@@ -493,75 +506,80 @@ export default (testContext: {
-        "message": {
-          "issuer": {
-            "id": "did:fake:123"
+        message: {
```
@lleifermann looks like there's some code duplication here causing test failures

lleifermann replied:
Fixed it. My prettier messed something up i believe.
